sweep/fee_bumper.go

```diff
@@ -0,0 +1,894 @@
+	//
+	// NOTE: The fee function is expected to have increased its returned
+	// fee rate after calling the SkipFeeBump method. So we can use it
+	// directly here.
+	tx, fee, err := t.createAndCheckTx(r.req, r.feeFunction)
+
+	// If the tx doesn't not have enought budget, we will return a result
+	// so the sweeper can handle it by re-clustering the utxos.
+	if errors.Is(err, ErrNotEnoughBudget) {
+		log.Warnf("Fail to fee bump tx %v: %v", oldTx.TxHash(), err)
+
+		return &BumpResult{
+			Event:     TxFailed,
+			Tx:        oldTx,
+			Err:       err,
+			requestID: requestID,
+		}, nil
+	}
+
+	// If the error is not budget related, we will return an error and let
+	// the fee bumper retry it at next block.
+	//
+	// NOTE: we can check the RBF error here and ask the fee function to
+	// recalculate the fee rate. However, this would defeat the purpose of
+	// using a deadline based fee function:
+	// - if the deadline is far away, there's no rush to RBF the tx.
+	// - if the deadline is close, we expect the fee function to give us a
+	//   higher fee rate. If the fee rate cannot satisfy the RBF rules, it
+	//   means the budget is not enough.
+	if err != nil {
+		log.Infof("Failed to bump tx %v: %v", oldTx.TxHash(), err)
+		return nil, err
+	}
+
+	// Register a new record by overwriting the same requestID.
+	t.records.Store(requestID, &monitorRecord{
+		tx:          tx,
+		req:         r.req,
+		feeFunction: r.feeFunction,
+		fee:         fee,
+	})
+
+	// Attempt to broadcast this new tx.
+	result, err := t.broadcast(requestID)
+	if err != nil {
+		return nil, err
+	}
+
+	// A successful replacement tx is created, attach the old tx.
+	result.ReplacedTx = oldTx
+
+	// If the new tx failed to be published, we will return the result so
+	// the caller can handle it.
+	if result.Event == TxFailed {
+		return result, nil
+	}
+
+	log.Infof("Replaced tx=%v with new tx=%v", oldTx.TxHash(), tx.TxHash())
+
+	// Otherwise, it's a successful RBF, set the event and return.
+	result.Event = TxReplaced
+	return result, nil
+}
+
+// calcCurrentConfTarget calculates the current confirmation target based on
+// the deadline height. The conf target is capped at 0 if the deadline has
+// already been past.
+func calcCurrentConfTarget(currentHeight, deadline int32) uint32 {
+	var confTarget uint32
+
+	// Calculate how many blocks left until the deadline.
+	deadlineDelta := deadline - currentHeight
+
+	// If we are already past the deadline, we will set the conf target to
+	// be 1.
+	if deadlineDelta <= 0 {
+		log.Warnf("Deadline is %d blocks behind current height %v",
+			-deadlineDelta, currentHeight)
+
+		confTarget = 1
+	} else {
+		confTarget = uint32(deadlineDelta)
+	}
+
+	return confTarget
+}
+
+// isConfirmed is a helper method that returns a boolean indicating whether the
+// tx is confirmed.
+func isConfirmed(event *chainntnfs.ConfirmationEvent) bool {
+	select {
+	// TODO(yy): check if there's a race between sending a block epoch and
```

Comment:
chatted offline, but leaving this note here: the block epoch ntfn is sent first and then the conf ntfn so yeah the conf ntfn could be missed

Reply:
switched to `GetTransactionDetails` instead.